pennylane/resource/resource.py

```diff
@@ -0,0 +1,169 @@
+        gate_types: {'Hadamard': 1, 'CNOT': 1}
+    """
+
+    def __init__(
+        self, num_wires=0, num_gates=0, gate_types={}, depth=0, shots=0
+    ):  # pylint: disable=too-many-arguments, dangerous-default-value
+        """Initialize a Resources instance and perform input type validation."""
+
+        if not all(isinstance(param, int) for param in [num_wires, num_gates, depth, shots]):
+            raise TypeError(
+                "Incorrect type of input, expected type int for num_wires, num_gates, depth and shots. "
+                f"Got {type(num_wires)}, {type(num_gates)}, {type(depth)}, {type(shots)} respectively. "
+            )
+
+        if not isinstance(gate_types, dict):
+            raise TypeError(
+                f"Incorrect type of input, expected type dict for gate_types. Got {type(gate_types)}."
+            )
+
+        if not all(val > -1 for val in [num_wires, num_gates, depth, shots]):
+            raise ValueError(
+                "Incorrect value of input, expected value of num_wires, num_gates, depth and shots to be >= 0."
+                f"Got {num_wires}, {num_gates}, {depth}, {shots} respectively. "
+            )
+
+        self._num_wires = num_wires
+        self._num_gates = num_gates
+        self._gate_types = gate_types
+        self._depth = depth
+        self._shots = shots
+
+    @property
+    def num_wires(self):
+        """Number of wires in the quantum workflow."""
+        return self._num_wires
+
+    @num_wires.setter
+    def num_wires(self, new_val):
+        """Setting is NOT allowed"""
+        raise AttributeError(
+            f"{type(self)} object does not support assignment for 'num_wires' attribute"
+        )
+
+    @property
+    def num_gates(self):
+        """Number of gates in the quantum workflow."""
+        return self._num_gates
+
+    @num_gates.setter
+    def num_gates(self, new_val):
+        """Setting is NOT allowed"""
+        raise AttributeError(
+            f"{type(self)} object does not support assignment for 'num_gates' attribute"
+        )
+
+    @property
+    def gate_types(self):
+        """Dictionary of unique quantum gates and counts in the quantum workflow."""
+        return self._gate_types
+
+    @gate_types.setter
+    def gate_types(self, new_val):
+        """Setting is NOT allowed"""
+        raise AttributeError(
+            f"{type(self)} object does not support assignment for 'gate_types' attribute"
+        )
+
+    @property
+    def depth(self):
+        """Circuit depth of the quantum workflow."""
+        return self._depth
+
+    @depth.setter
+    def depth(self, new_val):
+        """Setting is NOT allowed"""
+        raise AttributeError(
+            f"{type(self)} object does not support assignment for 'depth' attribute"
+        )
+
+    @property
+    def shots(self):
+        """Number of shots/samples in the quantum workflow."""
+        return self._shots
+
+    @shots.setter
+    def shots(self, new_val):
```

Comment:
Since setting attributes isn't allowed, you could set `Resources` to be a dataclass using the `@dataclass` decorator and set it to be frozen:

```python3
@dataclass(frozen=True)
class Resources:
    ...
 ```

Or, you could overwrite the `__setattr__` dunder method to raise an error. You can find an example [here](https://github.com/PennyLaneAI/pennylane/blob/c0fb9b851fcdf004b943f475cbeae47a86b7d7db/pennylane/measurements/shots.py#L130).

This way, you avoid having to create a setter for all the attributes and having a lot of repeated code.

Reply:
That was done initially, but the reason a frozen data class doesn't work is because the `gate_types` attribute must be a dictionary, which is mutable.

Reply:
With your current implementation, can't users still add/remove items from the dictionary, since you wouldn't change the reference to the dictionary by doing so?

Reply:
```pycon
>>> d = {"test": 1}
>>> res = Resources(gate_types=d)
>>> res
<Resource: wires=0, gates=0, depth=0, shots=0, gate_types={'test': 1}>
>>> d["new"] = 10
>>> d
{'test': 1, 'new': 10}
>>> res
<Resource: wires=0, gates=0, depth=0, shots=0, gate_types={'test': 1, 'new': 10}>
```